clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -903,6 +904,54 @@ private void updateBytesBuffered(boolean madeProgress) {
 
     @Override
     public long transferFrom(FileChannel fileChannel, long position, long count) throws IOException {
-        return fileChannel.transferTo(position, count, this);
+        if (state == State.CLOSING)
+            throw closingException();
+        if (state != State.READY)
+            return 0;
+
+        if (!flush(netWriteBuffer))
+            return 0;
+
+        long channelSize = fileChannel.size();
+        if (position > channelSize)
+            return 0;
+        int bytesToWrite = (int) Math.min(Math.min(count, channelSize - position), Integer.MAX_VALUE);
+
+        if (fileChannelBuffer == null) {
+            // same size as `netWriteBuffer`
+            int transferSize = 16384;
+            // Allocate a direct buffer to avoid one heap to heap buffer copy. SSLEngine copies the source
+            // buffer (fileChannelBuffer) to the destination buffer (netWriteBuffer) and then encrypts in-place.
+            // FileChannel.read() to a heap buffer requires a copy from a direct buffer to a heap buffer, which is not
+            // useful here.
+            fileChannelBuffer = ByteBuffer.allocateDirect(transferSize);
+        } else {
+            fileChannelBuffer.clear();
```

Comment:
We could consider using the data in this buffer due to the break after the `if (networkBytesWritten != bytesRead)` check or due to an exception. Generally, it seems like this should not happen if `netWriteBuffer` and `fileChannelBuffer` are the same size. It may still be worth doing that to be safe. @junrao, what do you think?

Reply:
We could reuse the remaining data in fileChannelBuffer, but those remaining bytes need to be included in bytesWritten so that the caller can issue the next transferFrom() from the right position.

Reply:
Yes, that's right. If you agree it's worth it, I'll tweak the code so that it doesn't discard the data in this case. I had added a log line when this happens, we can remove it if we make this change.

Reply:
Probably not worth it. The tricky thing is that we have to factor the remaining bytes in fileChannelBuffer into TransportLayers.hasPendingWrites(), which requires more work.

Reply:
I took a stab at doing this in a way where we don't have to take it into account in `hasPendingWrites`. It looks good to me, let me know what you think.